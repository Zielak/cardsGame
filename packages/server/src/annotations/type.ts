import {
  Context,
  DefinitionType,
  Schema,
  type as colType,
} from "@colyseus/schema"
import { globalContext } from "@colyseus/schema/lib/annotations"

import { globalEntitiesContext } from "./entitiesContext"

export function type(
  typeDef: DefinitionType,
  typesContext?: Context
): PropertyDecorator {
  return function (target: typeof Schema, field: string): void {
    const constructor = target.constructor as typeof Schema
    const context = globalEntitiesContext

    // let logType: any = `"${typeDef}"`
    // if (Array.isArray(typeDef)) {
    //   logType = `(array) [${typeDef[0]}]`
    // } else if (typeof typeDef === "function") {
    //   logType = `(${typeof typeDef}) "${typeDef.name}"`
    // } else if (typeof typeDef === "object") {
    //   logType = `(${typeof typeDef}) ${typeDef}`
    // }

    // Intercept type definition
    if (field.indexOf("children") !== 0) {
      if (!context.registeredTypeDefinitions.has(constructor.name)) {
        context.registeredTypeDefinitions.set(constructor.name, new Map())
      }
      // logs.verbose("child type:", `${constructor.name}.${field} ${logType}`)
      context.registeredTypeDefinitions
        .get(constructor.name)
        .set(field, typeDef)
    }

    console.log(`Type @${constructor.name}, field: ${field}`)
    return colType(typeDef, typesContext)(target, field)
  }
}

export function defineTypes(
  target: typeof Schema,
  fields: { [property: string]: DefinitionType },
  context?: Context // = globalContext
): typeof Schema {
  for (const field in fields) {
    type(fields[field], context)(target.prototype, field)
  }
  return target
}
