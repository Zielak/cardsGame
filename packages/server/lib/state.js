"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var colyseus_1 = require("colyseus");
var eventemitter3_1 = require("eventemitter3");
var logs_1 = require("./logs");
var utils_1 = require("@cardsgame/utils");
var entity_1 = require("./entity");
var entityMap_1 = require("./entityMap");
var State = /** @class */ (function (_super) {
    __extends(State, _super);
    function State(options) {
        var _this = _super.call(this) || this;
        // 60 cm
        _this.tableWidth = 600;
        // 60 cm
        _this.tableHeight = 600;
        _this.players = new entityMap_1.EntityMap();
        _this.isGameStarted = false;
        _this._lastID = -1;
        _this._allEntities = new Map();
        // TODO: do something with these options.
        _this.entities = new entity_1.Entity({
            state: _this,
            type: "root",
            name: "root"
        });
        _this.setupListeners();
        return _this;
    }
    Object.defineProperty(State.prototype, "currentPlayer", {
        get: function () {
            return this.players[this.currentPlayerIdx];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(State.prototype, "playersCount", {
        get: function () {
            return this.players.length;
        },
        enumerable: true,
        configurable: true
    });
    State.prototype.setupListeners = function () {
        var _this = this;
        this.on(utils_1.StateEvents.privatePropsSyncRequest, function (client) {
            // Bubble it down to every entity
            logs_1.logs.info("State.privatePropsSyncRequest");
            _this.entities.emit(utils_1.StateEvents.privatePropsSyncRequest, client);
        });
    };
    /**
     * FIXME: that should be client-side code...
     * Player playing the game needs to have his cards visible
     * at HIS pottom of the screen...
     */
    State.prototype.updatePlayers = function () {
        var _this = this;
        this.entities.childrenArray
            .filter(function (entity) { return entity.type === "player"; })
            .forEach(function (player, idx, players) {
            var angle = ((Math.PI * 2) / players.length) * idx;
            var point = {
                x: Math.sin(angle) * (_this.tableWidth * 0.4),
                y: Math.cos(angle) * (_this.tableHeight * 0.4)
            };
            player.angle = -utils_1.rad2deg(angle);
            player.x = point.x;
            player.y = point.y;
            logs_1.logs.log("updatePlayers", "player[" + idx + "] angle: " + angle + ", point: (" + point.x + ", " + point.y + ")");
        });
    };
    /**
     * Registers new entity to the gamestate
     * @param entity
     * @returns new ID to be assigned to that entity
     */
    State.prototype.rememberEntity = function (entity) {
        var newID = ++this._lastID;
        this._allEntities.set(newID, entity);
        return newID;
    };
    State.prototype.getEntity = function (idOrPath) {
        if (Array.isArray(idOrPath)) {
            var travel_1 = function (entity, path) {
                var idx = path.shift();
                var newChild = entity.children[idx];
                if (!newChild) {
                    logs_1.logs.error("getEntity/path", "This entity doesn't have such child");
                }
                if (path.length > 0) {
                    return travel_1(newChild, path);
                }
                else {
                    return newChild;
                }
            };
            return travel_1(this.entities, __spread(idOrPath));
        }
        return this._allEntities.get(idOrPath);
    };
    /**
     * Gets an array of all entities from the top-most parent
     * to the lowest of the child.
     */
    State.prototype.getEntitiesAlongPath = function (path) {
        var travel = function (entity, path, result) {
            if (result === void 0) { result = []; }
            var idx = path.shift();
            var newChild = entity.children[idx];
            if (!newChild) {
                logs_1.logs.error("getEntity/path", "This entity doesn't have such child");
            }
            result.push(newChild);
            if (path.length > 0) {
                return travel(newChild, path, result);
            }
            else {
                return result;
            }
        };
        return travel(this.entities, __spread(path));
    };
    State.prototype.logTreeState = function () {
        logs_1.logs.log("");
        var indent = function (level) {
            return "│ ".repeat(level);
        };
        var travel = function (entity, level) {
            if (level === void 0) { level = 0; }
            entity.childrenArray.map(function (child, idx, entities) {
                if (child.parentEntity.isContainer &&
                    entities.length > 5 &&
                    idx < entities.length - 5) {
                    // That's too much, man!
                    if (idx === 0) {
                        logs_1.logs.log("" + indent(level), "...");
                    }
                    return;
                }
                var owner = child.owner;
                var lastChild = entities.length - 1 === idx;
                var sIdx = idx === child.idx ? "" + idx : "e" + child.idx + ":s" + idx;
                var sVisibility = child.visibleToPublic ? "Pub" : "";
                var sChildren = child.length > 0 ? child.length + " children" : "";
                var sOwner = owner ? "(" + owner.type + " " + owner.clientID + ")" : "";
                var branchSymbol = lastChild ? "┕━" : "┝━";
                logs_1.logs.log("" + indent(level) + branchSymbol + "[" + sIdx + "]" + sVisibility, child.type + ":" + child.name + "-[" + child.idx + "]", sChildren, sOwner);
                if (child.length > 0) {
                    travel(child, level + 1);
                }
            });
        };
        var root = this.entities;
        logs_1.logs.log("\u250D\u2501[" + root.idx + "]", root.type + ":" + root.name, root.length + "children");
        travel(root, 1);
    };
    State.ROOT_ID = 0;
    __decorate([
        colyseus_1.nosync
    ], State.prototype, "_lastID", void 0);
    __decorate([
        colyseus_1.nosync
    ], State.prototype, "_allEntities", void 0);
    return State;
}(eventemitter3_1.EventEmitter));
exports.State = State;
// Get rid of EventEmitter stuff from the client
colyseus_1.nosync(State.prototype, "_events");
colyseus_1.nosync(State.prototype, "_eventsCount");
colyseus_1.nosync(State.prototype, "_maxListeners");
colyseus_1.nosync(State.prototype, "domain");
//# sourceMappingURL=state.js.map