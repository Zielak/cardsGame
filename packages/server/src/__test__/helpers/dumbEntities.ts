import { canBeChild } from "@/annotations/canBeChild.js"
import { containsChildren } from "@/annotations/containsChildren.js"
import { ChildTrait } from "@/traits/child.js"
import { applyTraitsMixins, Entity } from "@/traits/entity.js"
import { IdentityTrait } from "@/traits/identity.js"
import { ParentTrait } from "@/traits/parent.js"

@canBeChild
@containsChildren
@applyTraitsMixins([IdentityTrait, ParentTrait, ChildTrait])
export class DumbParent extends Entity<DumbParent> {}
export interface DumbParent extends IdentityTrait, ParentTrait, ChildTrait {}

@canBeChild
@applyTraitsMixins([IdentityTrait, ChildTrait])
export class DumbEntity extends Entity<DumbEntity> {}
export interface DumbEntity extends IdentityTrait, ChildTrait {}
