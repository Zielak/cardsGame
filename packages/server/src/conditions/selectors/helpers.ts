import { IEntity } from "../../traits/entity"
import { logs } from "../../logs"

export const propsMatch = (propName: string, values: any[]) => (
  entity: IEntity
) => {
  const result = values.some(value => entity[propName] === value)
  if (!result) {
    logs.warn(
      `propsMatch ${propName}`,
      `entity[${propName}] doesn't match any accepted values:`,
      values
    )
  }
  return result
}
