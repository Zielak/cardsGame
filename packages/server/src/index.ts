export { Server } from "colyseus"
export { RoomConstructor } from "colyseus/lib/Room"
export { Schema, ArraySchema, MapSchema } from "@colyseus/schema"

export { canBeChild } from "./annotations/canBeChild"
export { containsChildren } from "./annotations/containsChildren"
export { defineTypes, type } from "./annotations/type"

export * from "./bots/botNeuron"

import * as commands from "./commands"

export { commands }

export * from "./command"

export * from "./room"
export * from "./state/state"
export * from "./state/helpers"

export * from "./actionTemplate"

export { Conditions } from "./conditions"
export { getFlag, setFlag } from "./conditions/utils"
export { BotConditions, EntityConditions } from "./bots/conditions"
export { ClientEventConditions } from "./interaction"

export { testAction } from "./interaction"

export { Player, ServerPlayerEvent } from "./players/player"
export { Bot } from "./players/bot"

export * from "./traits/boxModel"
export * from "./traits/child"
export * from "./traits/entity"
export * from "./traits/flexyContainer"
export * from "./traits/identity"
export * from "./traits/label"
export * from "./traits/location"
export * from "./traits/ownership"
export * from "./traits/parent"
export * from "./traits/parentArray"
export * from "./traits/parentMap"
export * from "./traits/selectableChildren"
export * from "./traits/twoSided"

export * from "./entities"

export { populatePlayerEvent } from "./utils"
