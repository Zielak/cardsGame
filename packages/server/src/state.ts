import { nosync } from "colyseus"
import { EventEmitter } from "eventemitter3"
import { logs } from "./logs"
import { deg2rad, rad2deg, StateEvents } from "@cardsgame/utils"
import { Entity } from "./entity"
import { EntityMap } from "./entityMap"
import { Player } from "./player"

export class State extends EventEmitter {
  // 60 cm
  tableWidth = 600
  // 60 cm
  tableHeight = 600

  entities: Entity

  players = new EntityMap<PlayerData>()
  clients = new EntityMap<string>()
  currentPlayerIdx: number
  get currentPlayer(): PlayerData {
    return this.players[this.currentPlayerIdx]
  }
  get playersCount(): number {
    return this.players.length
  }

  // TODO: think this through:
  gameVariants: any

  isGameStarted = false
  @nosync
  _lastID = -1
  @nosync
  _allEntities = new Map<number, Entity>()

  constructor(options?: IStateOptions) {
    super()
    // TODO: do something with these options.

    this.entities = new Entity({
      state: this,
      type: "root",
      name: "root"
    })

    this.setupListeners()
  }

  setupListeners() {
    this.on(StateEvents.privatePropsSyncRequest, (client: string) => {
      // Bubble it down to every entity
      logs.info("State.privatePropsSyncRequest")
      this.entities.emit(StateEvents.privatePropsSyncRequest, client)
    })
  }

  /**
   * Registers new entity to the gamestate
   * @param entity
   * @returns new ID to be assigned to that entity
   */
  rememberEntity(entity: Entity) {
    const newID = ++this._lastID
    this._allEntities.set(newID, entity)
    return newID
  }

  /**
   * Get an Entity by its ID
   * @param id
   */
  getEntity(id: EntityID): Entity
  /**
   * Get an Entity by its idx path
   * @param path
   */
  getEntity(path: number[]): Entity
  getEntity(idOrPath: EntityID | number[]): Entity {
    if (Array.isArray(idOrPath)) {
      const travel = (entity: Entity, path: number[]) => {
        const idx = path.shift()
        const newChild = entity.children[idx] as Entity
        if (!newChild) {
          logs.error("getEntity/path", `This entity doesn't have such child`)
        }
        if (path.length > 0) {
          return travel(newChild, path)
        } else {
          return newChild
        }
      }
      return travel(this.entities, [...idOrPath])
    }
    return this._allEntities.get(idOrPath)
  }

  /**
   * Gets an array of all entities from the top-most parent
   * to the lowest of the child.
   */
  getEntitiesAlongPath(path: number[]): Entity[] {
    const travel = (entity: Entity, path: number[], result: Entity[] = []) => {
      const idx = path.shift()
      const newChild = entity.children[idx] as Entity
      if (!newChild) {
        logs.error("getEntity/path", `This entity doesn't have such child`)
      }
      result.push(newChild)
      if (path.length > 0) {
        return travel(newChild, path, result)
      } else {
        return result
      }
    }
    return travel(this.entities, [...path])
  }

  logTreeState() {
    logs.log("")
    const indent = (level: number) => {
      return "│ ".repeat(level)
    }
    const travel = (entity: Entity, level: number = 0) => {
      entity.childrenArray.map((child, idx, entities) => {
        if (
          child.parentEntity.isContainer &&
          entities.length > 5 &&
          idx < entities.length - 5
        ) {
          // That's too much, man!
          if (idx === 0) {
            logs.log(`${indent(level)}`, "...")
          }
          return
        }

        const owner = child.owner

        const lastChild = entities.length - 1 === idx
        const sIdx = idx === child.idx ? `${idx}` : `e${child.idx}:s${idx}`
        const sVisibility = child.visibleToPublic ? "Pub" : ""

        const sChildren = child.length > 0 ? child.length + " children" : ""
        const sOwner = owner ? `(${owner.type} ${owner.clientID})` : ""
        const branchSymbol = lastChild ? "┕━" : "┝━"

        logs.log(
          `${indent(level)}${branchSymbol}[${sIdx}]${sVisibility}`,
          `${child.type}:${child.name}-[${child.idx}]`,
          sChildren,
          sOwner
        )
        if (child.length > 0) {
          travel(child, level + 1)
        }
      })
    }
    const root = this.entities
    logs.log(
      `┍━[${root.idx}]`,
      `${root.type}:${root.name}`,
      root.length + "children"
    )
    travel(root, 1)
  }

  static ROOT_ID = 0
}

// Get rid of EventEmitter stuff from the client
nosync(State.prototype, "_events")
nosync(State.prototype, "_eventsCount")
nosync(State.prototype, "_maxListeners")
nosync(State.prototype, "domain")

export interface IStateOptions {
  minClients: number
  maxClients: number
  hostID: string
}

export type PlayerData = {
  entity: Entity
  clientID: string
}
