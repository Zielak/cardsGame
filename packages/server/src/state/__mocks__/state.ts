import type { Player } from "../../player/player.js"

export class State<
  V extends Record<string, unknown> = Record<string, unknown>
> {
  name = "Unnamed"
  type = "state"
  tableWidth = 60 // 60 cm
  tableHeight = 60 // 60 cm
  clients = new Array<string>()
  turnBased = true
  round = 0
  players = new Array<Player>()
  currentPlayerIdx = 0

  get currentPlayer(): Player {
    return this.turnBased ? this.players[this.currentPlayerIdx] : null
  }
  isGameStarted = false
  isGameOver = false

  ui = {}

  playerViewPosition = {
    alignX: "center",
    alignY: "bottom",
    paddingX: 0,
    paddingY: 0,
  }

  variantData: V
  hijacksInteractionTarget = false

  id = 0

  _registerEntity(): void {
    // do nothing
  }
}
