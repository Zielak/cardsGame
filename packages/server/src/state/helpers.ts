import { limit, logs } from "@cardsgame/utils"

import { Bot, isBot } from "../players/bot"
import { Player } from "../players/player"
import { IdentityTrait } from "../traits/identity"
import { hasChildren, isParent, ParentTrait } from "../traits/parent"

import { State } from "."

/**
 * Will get you an index of given player in turn queue.
 * Useful if you happen to have just a `Player` reference at hand.
 */
export function getPlayersIndex(state: State, player: Player): number {
  return state.players.findIndex((p) => p === player)
}

export function getNextPlayerIdx(state: State): number {
  const current = limit(state.currentPlayerIdx, 0, state.players.length - 1)
  return current + 1 === state.players.length ? 0 : current + 1
}

export function getNextPlayer(state: State): Player {
  return state.players[getNextPlayerIdx(state)]
}

export function getPreviousPlayerIdx(state: State): number {
  const current = limit(state.currentPlayerIdx, 0, state.players.length - 1)
  return current - 1 === -1 ? state.players.length - 1 : current - 1
}

export function getPreviousPlayer(state: State): Player {
  return state.players[getPreviousPlayerIdx(state)]
}

export function getPlayerByName(state: State, name: string): Player {
  return state.players.find((player) => player.name === name)
}

export function getAllBots(state: State): Bot[] {
  return state.players.filter((player) => isBot(player)) as Bot[]
}

/**
 * Gets an array of all entities from the top-most parent
 * to the lowest of the child.
 */
export function getEntitiesAlongPath(
  state: State,
  path: number[]
): IdentityTrait[] {
  const travel = (
    parent: ParentTrait,
    remainingPath: number[],
    result: any[] = []
  ): any[] => {
    const idx = remainingPath.shift()
    const newChild = parent.getChild(idx)
    if (!newChild) {
      logs.error("getEntitiesAlongPath", `This entity doesn't have such child.`)
      return result
    }

    if (remainingPath.length > 0 && !hasChildren(newChild)) {
      logs.error(
        "getEntitiesAlongPath",
        `Path inaccessible, entity doesn't have any children. Stopped at [${path}]. Remaining path: [${remainingPath}]`
      )
      return result
    }

    result.push(newChild)
    if (remainingPath.length > 0 && isParent(newChild)) {
      return travel(newChild, remainingPath, result)
    } else {
      return result
    }
  }
  return travel(state, [...path])
}
