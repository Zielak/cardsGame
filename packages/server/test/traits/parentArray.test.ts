import { State } from "src/state"
import type { ChildTrait } from "src/traits/child"
import type { IdentityTrait } from "src/traits/identity"

import { DumbEntity, DumbParent } from "../helpers/dumbEntities"

let state: State

beforeEach(() => {
  state = new State()
})

describe("ParentConstructor", () => {
  test("sets default properties", () => {
    const parent = new DumbParent(state)

    expect(parent.countChildren()).toBe(0)

    expect(parent.hijacksInteractionTarget).toBeDefined()
    expect(parent.hijacksInteractionTarget).toBe(true)
  })
})

describe("#removeChildAt", () => {
  let parent: DumbParent
  beforeEach(() => {
    parent = new DumbParent(state)
  })
  it(`throws on invalid "idx"`, () => {
    expect(parent.removeChildAt(-2)).toBe(false)
    expect(parent.removeChildAt(NaN)).toBe(false)
    expect(parent.removeChildAt(Infinity)).toBe(false)
    expect(parent.removeChildAt(10)).toBe(false)
  })
  it("removes the only child", () => {
    new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(1)
    expect(parent.removeChildAt(0)).toBe(true)
    expect(parent.countChildren()).toBe(0)
  })
  it("removes child in the middle", () => {
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(3)
    expect(parent.removeChildAt(1)).toBe(true)
    expect(parent.countChildren()).toBe(2)
  })
  it("returns false for not its child", () => {
    const child = new DumbEntity(state)

    expect(parent.removeChild(child)).toBe(false)
  })
})

describe("#removeChild", () => {
  let parent: DumbParent
  it("calls removeChildAt", () => {
    parent = new DumbParent(state)
    const entity = new DumbEntity(state, { parent })

    expect(parent.countChildren()).toBe(1)
    expect(parent.removeChild(entity)).toBe(true)
    expect(parent.countChildren()).toBe(0)
  })

  it("finds proper child", () => {
    parent = new DumbParent(state)
    new DumbEntity(state, { parent })
    const entity = new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })

    parent.removeChild(entity)
    expect(
      parent.getChildren().map((e: ChildTrait & IdentityTrait) => e.id)
    ).toMatchSnapshot()
  })
})

describe("#addChild", () => {
  let parent: DumbParent
  let entity: DumbEntity
  beforeEach(() => {
    parent = new DumbParent(state)
    new DumbEntity(state, { parent })
  })
  it("adds new item", () => {
    entity = new DumbEntity(state)
    parent.addChild(entity)

    expect(entity.parent).toBe(parent)
    expect(entity.idx).toBe(1)
    expect(parent.countChildren()).toBe(2)
  })
  it("prepends new item", () => {
    entity = new DumbEntity(state)
    parent.addChild(entity, true)

    expect(entity.parent).toBe(parent)
    expect(entity.idx).toBe(0)
    expect(parent.countChildren()).toBe(2)
  })
  it("add item at index", () => {
    const entityA = new DumbEntity(state)
    const entityB = new DumbEntity(state)

    parent.addChild(entityA, 0)

    expect(entityA.parent).toBe(parent)
    expect(entityA.idx).toBe(0)
    expect(parent.countChildren()).toBe(2)

    parent.addChild(entityB, 1)

    expect(entityB.parent).toBe(parent)
    expect(entityB.idx).toBe(1)
    expect(entityA.idx).toBe(0)
    expect(parent.countChildren()).toBe(3)
  })
  it("throws on incorrect indexes", () => {
    const e = new DumbEntity(state)
    expect(() => parent.addChild(e, -1)).toThrow()
  })
  it("throws on empty argument", () => {
    expect(() => {
      // @ts-expect-error: testing empty argumentgetChild
      parent.addChild()
    }).toThrow(`missing required argument "entity"`)
  })
})

describe("#moveChildTo", () => {
  const mapArrayChildren = (p): any =>
    p.getChildren().map((e) => (e ? e.id : null))

  let parent: DumbParent
  beforeEach(() => {
    parent = new DumbParent(state)
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
  })
  it("moves item3 down 1 position", () => {
    parent.moveChildTo(2, 1)
    expect(mapArrayChildren(parent)).toMatchSnapshot()
  })
  it("moves item3 up 1 position", () => {
    parent.moveChildTo(2, 3)
    expect(mapArrayChildren(parent)).toMatchSnapshot()
  })
  it("moves item3 to bottom", () => {
    parent.moveChildTo(2, 0)
    expect(mapArrayChildren(parent)).toMatchSnapshot()
  })
  it("moves item3 to top", () => {
    parent.moveChildTo(2, parent.countChildren() - 1)
    expect(mapArrayChildren(parent)).toMatchSnapshot()
  })
  it("doesn't move anything (item3 to same position)", () => {
    parent.moveChildTo(2, 2)
    expect(mapArrayChildren(parent)).toMatchSnapshot()
  })
  it("handles too high 'to' number, treats as max value", () => {
    expect(() => parent.moveChildTo(2, 10)).not.toThrow()
    expect(mapArrayChildren(parent)).toMatchSnapshot()
  })

  it("throws on out of range idx", () => {
    expect(() => parent.moveChildTo(-2, 3)).toThrow()
  })
})

test.todo("#countChildren")
describe("#getChildren", () => {
  let parent: DumbParent
  beforeEach(() => {
    parent = new DumbParent(state)
  })
  test("empty parent", () => {
    expect(parent.getChildren()).toMatchSnapshot()
  })
  test("couple of children", () => {
    const first = new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    const last = new DumbEntity(state, { parent })

    const result = parent.getChildren()
    expect(result).toMatchSnapshot()
    expect(result[0]).toBe(first)
    expect(result[result.length - 1]).toBe(last)
  })
  it("doesn't return direct reference", () => {
    expect(parent.getChildren() === parent["allChildren"]).toBeFalsy()
  })
})

describe("#getChild", () => {
  describe("ArrayParent", () => {
    let parent
    let entity

    beforeEach(() => {
      parent = new DumbParent(state)
    })
    it("gets child", () => {
      new DumbEntity(state, { parent })
      entity = new DumbEntity(state, { parent })

      expect(parent.getChild(1)).toBe(entity)
    })
    it.todo("gets nothing")
    it.todo("throws if idx out of range")
  })
})

describe("#getTop", () => {
  test("ArrayParent", () => {
    const parent = new DumbParent(state)
    new DumbEntity(state, { parent })
    new DumbEntity(state, { parent })
    const entity = new DumbEntity(state, { parent })

    expect(parent.getTop()).toBe(entity)
  })
})

test("#getBottom", () => {
  const parent = new DumbParent(state)
  const entity = new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })
  new DumbEntity(state, { parent })

  expect(parent.getBottom()).toBe(entity)
})

describe("#indexFits", () => {
  let parent: DumbParent
  describe("limitless", () => {
    beforeEach(() => {
      parent = new DumbParent(state)
    })
    test("empty container", () => {
      expect(parent.indexFits(0)).toBe(true)
      expect(parent.indexFits(1)).toBe(true)
      expect(parent.indexFits(100)).toBe(true)

      expect(parent.indexFits(-1)).toBe(false)
    })
    test("containing some children", () => {
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })

      expect(parent.indexFits(0)).toBe(true)
      expect(parent.indexFits(1)).toBe(true)
      expect(parent.indexFits(100)).toBe(true)

      expect(parent.indexFits(-1)).toBe(false)
    })
  })

  describe("maxChildren", () => {
    beforeEach(() => {
      parent = new DumbParent(state, { maxChildren: 2 })
    })
    test("empty container", () => {
      expect(parent.indexFits(0)).toBe(true)
      expect(parent.indexFits(1)).toBe(true)

      expect(parent.indexFits(100)).toBe(false)
      expect(parent.indexFits(-1)).toBe(false)
    })
    it("shouldn't be checking if spots are occupied", () => {
      new DumbEntity(state, { parent })
      new DumbEntity(state, { parent })

      expect(parent.indexFits(0)).toBe(true)
      expect(parent.indexFits(1)).toBe(true)

      expect(parent.indexFits(100)).toBe(false)
      expect(parent.indexFits(-1)).toBe(false)
    })
  })
})
