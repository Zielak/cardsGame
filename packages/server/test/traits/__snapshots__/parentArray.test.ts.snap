// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getChildren couple of children 1`] = `
Array [
  Object {
    "idx": 0,
  },
  Object {
    "idx": 1,
  },
  Object {
    "idx": 2,
  },
  Object {
    "idx": 3,
  },
  Object {
    "idx": 4,
  },
]
`;

exports[`#getChildren empty parent 1`] = `Array []`;

exports[`#moveChildTo doesn't move anything (item3 to same position) 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`#moveChildTo handles too high 'to' number, treats as max value 1`] = `
Array [
  1,
  2,
  4,
  5,
  3,
]
`;

exports[`#moveChildTo moves item3 down 1 position 1`] = `
Array [
  1,
  3,
  2,
  4,
  5,
]
`;

exports[`#moveChildTo moves item3 to bottom 1`] = `
Array [
  3,
  1,
  2,
  4,
  5,
]
`;

exports[`#moveChildTo moves item3 to top 1`] = `
Array [
  1,
  2,
  4,
  5,
  3,
]
`;

exports[`#moveChildTo moves item3 up 1 position 1`] = `
Array [
  1,
  2,
  4,
  3,
  5,
]
`;

exports[`#removeChild finds proper child 1`] = `
Array [
  1,
  3,
]
`;
